README
Win32-IPC version 1.07, released April 16, 2008

The Win32-IPC modules provide access to the Win32 sychronization
objects, which are (currently) documented at

  http://msdn2.microsoft.com/en-us/library/ms686364(VS.85).aspx

The supported sychronization objects and associated modules are:
  Event     - Win32::Event
  Mutex     - Win32::Mutex
  Semaphore - Win32::Semaphore

In addition, the Win32::ChangeNotify module provides access to
directory change notifications, which let you monitor a specified
directory tree for file modifications.  While not strictly a
synchronization object, change notifications are often used in similar
ways.



INSTALLATION

To install these modules, run the following commands:

	perl Build.PL
	Build
	Build test
	Build install



DEPENDENCIES

You'll need a suitable C compiler.
Microsoft Visual C++ 6 or MinGW GCC should work.
You can also build it with GCC under Cygwin.



CHANGES
    Here's what's new in version 1.07 of Win32-IPC:
    (See the file "Changes" for the full revision history.)

	- There are no significant functional changes in 1.07.	If
	  you've already installed 1.06, don't bother upgrading.

	- Give a more useful error message if Module::Build is not installed

	- Don't bother to test $! after fetching constant('INFINITE')

	- Added LICENSE AND COPYRIGHT section to documentation

1.06  April 15, 2008
	- *** INCOMPATIBLE CHANGE ***

	  Passing undef as the value of an optional parameter is now
	  the same as omitting that parameter.	In previous versions,
	  undef was interpreted as either the empty string or 0 (along
	  with a warning about "Use of uninitialized value...").

	  In particular, if you pass undef as the timeout, it now
	  means INFINITE, but to previous versions it meant 0.

	- The modules now use XSLoader instead of DynaLoader.

	- Copied Yves Orton's fix (from Win32::ChangeNotify 1.04) so
	  "a true value" no longer has to be a non-zero number to all
	  relevant places.

	- Win32::IPC and its subclasses are now distributed separately
	  from libwin32 (which has been broken up).  Prior to this
	  release, each module had its own Changes file, and their
	  version numbers were not synchronized.  Those files are
	  appended below.



COPYRIGHT AND LICENSE

Copyright 1998-2008 Christopher J. Madsen

 Created: 3 Feb 1998 from the ActiveWare version
   (c) 1995 Microsoft Corporation. All rights reserved.
       Developed by ActiveWare Internet Corp., http://www.ActiveState.com

   Other modifications (c) 1997 by Gurusamy Sarathy <gsar@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for the Perl Win32-IPC modules.

1.07  April 16, 2008
	- There are no significant functional changes in 1.07.	If
	  you've already installed 1.06, don't bother upgrading.

	- Give a more useful error message if Module::Build is not installed

	- Don't bother to test $! after fetching constant('INFINITE')

	- Added LICENSE AND COPYRIGHT section to documentation

1.06  April 15, 2008
	- *** INCOMPATIBLE CHANGE ***

	  Passing undef as the value of an optional parameter is now
	  the same as omitting that parameter.	In previous versions,
	  undef was interpreted as either the empty string or 0 (along
	  with a warning about "Use of uninitialized value...").

	  In particular, if you pass undef as the timeout, it now
	  means INFINITE, but to previous versions it meant 0.

	- The modules now use XSLoader instead of DynaLoader.

	- Copied Yves Orton's fix (from Win32::ChangeNotify 1.04) so
	  "a true value" no longer has to be a non-zero number to all
	  relevant places.

	- Win32::IPC and its subclasses are now distributed separately
	  from libwin32 (which has been broken up).  Prior to this
	  release, each module had its own Changes file, and their
	  version numbers were not synchronized.  Those files are
	  appended below.


======================================================================
Revision history for Perl extension Win32::IPC.

1.03  Fri Jul 11 15:40:00 2003
	- revised how WaitForMultiple handles non-Win32::IPC objects

1.02  Mon May 22 21:16:37 2000
	- support for building under Perl v5.6.0

1.01  Tue Sep 15 12:25:16 1998
	- use New/Safefree instead of safemalloc/safefree

1.00  Fri Feb  6 1998 10:51:40
	- Rewritten from ActiveWare version by Christopher J. Madsen
	- Simplified code; converted from C++ to C
	- Fixed fatal bugs in WaitForMultipleObjects
	- New interface using lower-case method names

0.03  Mon Feb  2 13:44:17 1998
	- added Win32::Event to WaitForMultipleObjects

0.02  Fri Jul  4 12:19:12 1997
	- fixed refs to obsolete "ChangeNotification"

0.01  Sat Apr  5 14:24:55 1997
	- original version; created by h2xs 1.18


======================================================================
Revision history for Perl extension Win32::ChangeNotify.

1.05  2005/09/06
	- Remove USING_WIDE() code and fix cygwin compiler warnings

1.04  2002/12/18 21:48
	- Fixed $subtree handling. Up till now it had to be a true numeric value
	  contrary to what the docs say "a true value...." (Yves Orton)

1.03  Mon May 22 2000 20:32:04
        - support for passing Unicode strings to new() method

1.02  Sun Jun 13 1999 12:33:09
	- Bug fix to AUTOLOAD (FILE_NOTIFY_CHANGE_* constants now work)
	  (courtesy Christopher J. Madsen)

1.01  Sun Apr 26 1998 13:28:49
	- Bug fixes to `new' and `close' methods (courtesy
	  Chris Madsen)

1.00  Fri Feb  6 1998 10:35:22
	- Rewritten from ActiveWare version by Christopher J. Madsen
	- Simplified code; converted from C++ to C
	- New interface using lower-case method names

0.01  Sat Apr  5 1997 14:24:50
	- original version; created by h2xs 1.18


======================================================================
Revision history for Perl extension Win32::Event.

1.01  Mon May 22 2000 20:56:41
	- support for passing Unicode strings to the new() and open() methods
	  under Perl 5.6.0 (thanks to Doug Lankshear <doug@lankshear.net>)

1.00  Fri Feb  6 1998 10:35:22
	- Rewritten from ActiveWare version by Christopher J. Madsen
	- Simplified code; converted from C++ to C
	- New interface using lower-case method names

0.01  Mon Feb 2 1998 12:39:07
	- original version; copied from Win32::Mutex 0.01 (libwin32 0.9)


======================================================================
Revision history for Perl extension Win32::Mutex.

1.02  Mon May 22 21:19:43 2000
	- support for passing Unicode strings to the new() and open()
	  methods (thanks to Doug Lankshear <doug@lankshear.net>)

1.01  Sun Feb  8 1998 20:17:15
	- Added Open() function for compat

1.00  Fri Feb  6 1998 10:56:09
	- Rewritten from ActiveWare version by Christopher J. Madsen
	- Simplified code; converted from C++ to C
	- New interface using lower-case method names

0.01  Sat Apr  5 14:25:17 1997
	- original version; created by h2xs 1.18


======================================================================
Revision history for Perl extension Win32::Semaphore.

1.02  Mon May 22 2000 21:53:04
	- support for passing Unicode strings to new() and open() methods
	  (Doug Lankshear <doug@lankshear.net>)

1.01  Sun Feb  8 1998 20:16:56
	- Add Open() function for compat

1.00  Fri Feb  6 1998 11:01:57
	- Rewritten from ActiveWare version by Christopher J. Madsen
	- Simplified code; converted from C++ to C
	- New interface using lower-case method names
	- Created some tests in test.pl

0.01  Sat Apr  5 14:24:13 1997
	- original version; created by h2xs 1.18
	- imported Activeware version


lib/Win32/ChangeNotify.pm
#---------------------------------------------------------------------
package Win32::ChangeNotify;
#
# Copyright 1998-2008 Christopher J. Madsen
#
# Created: 3 Feb 1998 from the ActiveWare version
#   (c) 1995 Microsoft Corporation. All rights reserved.
#       Developed by ActiveWare Internet Corp., http://www.ActiveState.com
#
#   Other modifications (c) 1997 by Gurusamy Sarathy <gsar@cpan.org>
#
# Author: Christopher J. Madsen <perl@cjmweb.net>
# $Id: ChangeNotify.pm 256 2008-03-02 03:07:35Z cmadsn $
#
# This program is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either the
# GNU General Public License or the Artistic License for more details.
#
# Monitor directory for changes
#---------------------------------------------------------------------

use strict;
use warnings;
use vars qw($AUTOLOAD $VERSION @ISA @EXPORT @EXPORT_OK);

use Carp;
use Win32::IPC 1.00 '/./';      # Import everything

BEGIN
{
  $VERSION = '1.06';

  @ISA = qw(Win32::IPC);        # Win32::IPC isa Exporter
  @EXPORT = qw(
    FILE_NOTIFY_CHANGE_ATTRIBUTES
    FILE_NOTIFY_CHANGE_DIR_NAME
    FILE_NOTIFY_CHANGE_FILE_NAME
    FILE_NOTIFY_CHANGE_LAST_WRITE
    FILE_NOTIFY_CHANGE_SECURITY
    FILE_NOTIFY_CHANGE_SIZE
    INFINITE
  );
  @EXPORT_OK = qw(
    wait_any wait_all
  );

  require XSLoader;
  XSLoader::load('Win32::ChangeNotify', $VERSION);
} # end BEGIN bootstrap

sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.

    my $constname;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    if ($constname =~ /^FILE_NOTIFY_CHANGE_/) {
	local $! = 0;
        my $val = constant($constname);
        croak("$constname is not defined by Win32::ChangeNotify") if $! != 0;
        eval "sub $AUTOLOAD () { $val }";
        goto &$AUTOLOAD;
    }
} # end AUTOLOAD

sub new {
    my ($class,$path,$subtree,$filter) = @_;

    if ($filter =~ /\A[\s|A-Z_]+\Z/i) {
        $filter = 0;
        foreach (split(/[\s|]+/, $_[3])) {
            $filter |= constant("FILE_NOTIFY_CHANGE_" . uc $_);
            carp "Invalid filter $_" if $!;
        }
    }
    _new($class,$path,$subtree,$filter);
} # end new

# Deprecated ActiveWare functions:
sub FindFirst { $_[0] = Win32::ChangeNotify->_new(@_[1..3]); }
*Close    = \&close;            # Alias close to Close
*FindNext = \&reset;            # Alias reset to FindNext

1;
__END__

=head1 NAME

Win32::ChangeNotify - Monitor events related to files and directories

=head1 VERSION

This document describes version 1.06 of Win32::ChangeNotify, released April 15, 2008 as part of Win32-IPC version 1.06.


=head1 SYNOPSIS

	require Win32::ChangeNotify;

	$notify = Win32::ChangeNotify->new($Path,$WatchSubTree,$Events);
	$notify->wait or warn "Something failed: $!\n";
	# There has been a change.

=head1 DESCRIPTION

This module allows the user to use a Win32 change notification event
object from Perl.  This allows the Perl program to monitor events
relating to files and directory trees.

Unfortunately, the Win32 API which implements this feature does not
provide any indication of I<what> triggered the notification (as far
as I know).  If you're monitoring a directory for file changes, and
you need to know I<which> file changed, you'll have to find some other
way of determining that.  Depending on exactly what you're trying to
do, you may be able to check file timestamps to find recently changed
files.  Or, you may need to cache the directory contents somewhere and
compare the current contents to your cached copy when you receive a
change notification.

The C<wait> method and C<wait_all> & C<wait_any> functions are
inherited from the L<Win32::IPC> module.

=head2 Methods

=over 4

=item $notify = Win32::ChangeNotify->new($path, $subtree, $filter)

Constructor for a new ChangeNotification object.  C<$path> is the
directory to monitor.  If C<$subtree> is true, then all directories
under C<$path> will be monitored.  C<$filter> indicates what events
should trigger a notification.  It should be a string containing any
of the following flags (separated by whitespace and/or C<|>).

   ATTRIBUTES	Any attribute change
   DIR_NAME     Any directory name change
   FILE_NAME    Any file name change (creating/deleting/renaming)
   LAST_WRITE   Any change to a file's last write time
   SECURITY     Any security descriptor change
   SIZE         Any change in a file's size

(C<$filter> can also be an integer composed from the
C<FILE_NOTIFY_CHANGE_*> constants.)

Under Cygwin, C<$path> must be a Windows pathname, not a Cygwin
Unix-like pathname.

=item $notify->close

Shut down monitoring.  You could just C<undef $notify> instead (but
C<close> works even if there are other copies of the object).  This
happens automatically when your program exits.

=item $notify->reset

Resets the ChangeNotification object after a change has been detected.
The object will become signalled again after the next change.  (It is
OK to call this immediately after C<new>, but it is not required.)
Returns true if successful, or zero if it fails (additional error
information can be found in C<$^E>).

=item $notify->wait

See L<Win32::IPC>.  Remember to call C<reset> afterwards if you want
to continue monitoring.

=back

=head2 Deprecated Functions and Methods

B<Win32::ChangeNotify> still supports the ActiveWare syntax, but its
use is deprecated.

=over 4

=item FindFirst($Obj,$PathName,$WatchSubTree,$Filter)

Use

  $Obj = Win32::ChangeNotify->new($PathName,$WatchSubTree,$Filter)

instead.

=item $obj->FindNext()

Use C<$obj-E<gt>reset> instead.

=item $obj->Close()

Use C<$obj-E<gt>close> instead.

=back


=head1 DIAGNOSTICS

None.


=head1 CONFIGURATION AND ENVIRONMENT

Win32::ChangeNotify requires no configuration files or environment variables.

It runs under 32-bit or 64-bit Microsoft Windows, either natively or
under Cygwin.


=head1 DEPENDENCIES

L<Win32::IPC>


=head1 INCOMPATIBILITIES

Prior to version 1.06, Win32::ChangeNotify treated C<undef> values
differently.  In version 1.06 and later, passing C<undef> as the value
of an optional parameter is the same as omitting that parameter.  In
previous versions, C<undef> was interpreted as either the empty string
or 0 (along with a warning about "Use of uninitialized value...").


=head1 BUGS AND LIMITATIONS

No bugs have been reported.


=head1 AUTHOR

Christopher J. Madsen E<lt>F<perl AT cjmweb.net>E<gt>

Please report any bugs or feature requests to
S<< C<< <bug-Win32-IPC AT rt.cpan.org> >> >>,
or through the web interface at
L<http://rt.cpan.org/Public/Bug/Report.html?Queue=Win32-IPC>

Loosely based on the original module by ActiveWare Internet Corp.,
L<http://www.ActiveState.com>

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENSE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.

=cut

# Local Variables:
# tmtrack-file-task: "Win32::ChangeNotify"
# End:
